### PR TITLE
adding image caption support

### PR DESCRIPTION
**Pull request by msxingpeng**

### CODE REVIEW

File: vision_datasets/common/constants.py [hunk 1]
```diff
@@ -20,7 +21,7 @@ class Usages:
 class Formats:
     IRIS = 'iris'
     COCO = 'coco'
-
+    IMCAP = 'imcap'
```
Comment (pj-ms) [inline vision_datasets/common/constants.py]: can you run a flake8 check? Thx

File: vision_datasets/common/data_manifest.py [hunk 2]
```diff
@@ -470,3 +471,62 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
...
+                parts = line.rsplit(' ', maxsplit=1)  # assumption: only the image file path can have spaces
+                img_path = parts[0]
+                image_id = parts[1]
+                w, h = img_wh[img_path] if img_wh else (None, None)
+                image_id = int(image_id)
+                images.append(ImageDataManifest(img_path, get_full_sas_or_path(img_path), w, h, image_id))
+            file_reader.close()
+        return DatasetManifest(images, None, dataset_info.type)
+
+    @staticmethod
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py]: avoid copy-paste

File: vision_datasets/common/constants.py (hunk 1 above)
Comment (pj-ms) [inline vision_datasets/common/constants.py]: We have a special format for captioning task which is not irispy or coco?

File: vision_datasets/common/data_manifest.py [hunk 3]
```diff
@@ -470,3 +471,62 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
         images.sort(key=lambda x: x.id)
 
         return DatasetManifest(images, labelmap, data_type)
+
+
+class ImCapManifestAdaptor:
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py]: need documentation.

**Comment by pj-ms** [comment]
Missing unit tests

File: README.md [hunk 4]
```diff
@@ -109,6 +109,31 @@ Note that
 - Note that we used to use `ltrb` as default. If your coco annotations were prepared to work with this repo before version 0.1.2. Please add `"bbox_format": "ltrb"` to your coco file.
 - Regardless of what format bboxes are stored in Coco file, when annotations are transformed into `ImageDataManifest`, the bbox will be unified into `ltrb: [left, top, right, bottom]`.
 
+##### Coco JSON - Image caption
+
+Here is one example of the json file for image image caption task.
```
Comment (pj-ms) [inline README.md]: nit: image image => image

File: vision_datasets/common/data_manifest.py [hunk 5]
```diff
@@ -447,6 +451,17 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
 
         file_reader.close()
 
+        if data_type == DatasetTypes.IMCAP:
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:534]: Can you add a check in IrisManifestAdaptor to reject this type?

File: vision_datasets/common/data_manifest.py [hunk 6]
```diff
@@ -447,6 +451,17 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
...
+            images_by_id = {}
+            for img in coco_manifest['images']:
+                label = {'image_id': img['id'], 'file_name': img['file_name'], 'caption': []}
+                images_by_id.update({img['id']: ImageDataManifest(img['id'], get_full_sas_or_path(img['file_name']), [], [], label)})
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py]: why pass [] and [] for image width and height?

File: vision_datasets/common/data_manifest.py [hunk 7]
```diff
@@ -447,6 +451,17 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
 
         file_reader.close()
 
+        if data_type == DatasetTypes.IMCAP:
+            images_by_id = {}
+            for img in coco_manifest['images']:
+                label = {'image_id': img['id'], 'file_name': img['file_name'], 'caption': []}
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py]: why do we want to save the image id and file_name again in the label?

File: vision_datasets/common/data_manifest.py (hunk 6 above)
Comment (msxingpeng) [reply vision_datasets/common/data_manifest.py]: The ground truth file and the prediction file for image caption contain no image width and height.

File: vision_datasets/common/data_manifest.py (hunk 7 above)
Comment (msxingpeng) [reply vision_datasets/common/data_manifest.py]: This 'image_id' and 'file_name' information is used to construct the prediction json file in the inference step, so that we can use pycocoevalcap to evaluate the result

File: vision_datasets/common/data_manifest.py (hunk 5 above)
Comment (msxingpeng) [reply vision_datasets/common/data_manifest.py:534]: Done!

File: README.md (hunk 4 above)
Comment (msxingpeng) [reply README.md]: fixed

File: vision_datasets/common/data_manifest.py (hunk 7 above)
Comment (pj-ms) [reply vision_datasets/common/data_manifest.py]: ImageDataManifest has those info. why save it again in the label?

File: vision_datasets/common/data_manifest.py (hunk 6 above)
Comment (pj-ms) [reply vision_datasets/common/data_manifest.py]: still not justifying passing [] somewhere expecting int....

File: tests/test_create_dataset_manifest.py [hunk 8]
```diff
@@ -300,6 +300,48 @@ def test_object_detection_bbox_format_LTRB(self):
...
+                {"id": 3, "image_id": 1, "caption": "A black Honda motorcycle with a dark burgundy seat."},
+                {"id": 4, "image_id": 1, "caption": "Ma motorcycle parked on the gravel in front of a garage."},
+                {"id": 5, "image_id": 1, "caption": "A motorcycle with its brake extended standing outside."},
+                {"id": 6, "image_id": 2, "caption": "A picture of a modern looking kitchen area.\n"},
+                {"id": 7, "image_id": 2, "caption": "A narrow kitchen ending with a chrome refrigerator."},
+                {"id": 8, "image_id": 2, "caption": "A narrow kitchen is decorated in shades of white, gray, and black."},
+                {"id": 9, "image_id": 2, "caption": "a room that has a stove and a icebox in it"},
+                {"id": 10, "image_id": 2, "caption": "A long empty, minimal modern skylit home kitchen."}
+            ],
+            "type": "captions"
```
Comment (pj-ms) [inline tests/test_create_dataset_manifest.py]: image_caption or captions? Is this test working? Thanks

File: vision_datasets/common/data_manifest.py [hunk 9]
```diff
@@ -325,11 +329,11 @@ def create_dataset_manifest(dataset_info, usage: str, container_sas_or_root_dir:
         """
         assert dataset_info
         assert usage
-
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py:407]: nit: having an empty line after assert is good for readability

File: vision_datasets/common/data_manifest.py [hunk 10]
```diff
@@ -91,7 +91,11 @@ def __init__(self, id, img_path, width, height, labels):
             img_path (str): path to image
             width (int): image width
             height (int): image height
-            labels (list or dict): classification: [c_id] for multiclass, [c_id1, c_id2, ...] for multilabel; detection: [c_id, left, top, right, bottom]; dict[task, labels] for multitask dataset
+            labels (list or dict):
+                classification: [c_id] for multiclass, [c_id1, c_id2, ...] for multilabel;
+                detection: [c_id, left, top, right, bottom];
+                image caption: dict of {'image_id': image_id, 'image_file': filename, 'caption': [dummy strings]};
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py]: Does that mean one image can have multiple captions?

Comment (pj-ms) [reply vision_datasets/common/data_manifest.py]: Seems not the case in the unit tests

File: tests/test_create_dataset_manifest.py (hunk 8 above)
Comment (msxingpeng) [reply tests/test_create_dataset_manifest.py]: The key "type" is not used in the dataloader, so it can be removed. I have removed it

File: vision_datasets/common/data_manifest.py (hunk 10 above)
Comment (msxingpeng) [reply vision_datasets/common/data_manifest.py]: Yes, one image can have multiple captions. In the unit tests, each image has 5 captions.

Comment (pj-ms) [reply vision_datasets/common/data_manifest.py]: got it.

File: vision_datasets/common/data_manifest.py [hunk 11]
```diff
@@ -447,6 +452,14 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
 
         file_reader.close()
 
+        if data_type == DatasetTypes.IMCAP:
+            images_by_id = {img['id']: ImageDataManifest(img['id'], get_full_sas_or_path(img['file_name']), None, None, []) for img in coco_manifest['images']}
+            for annotation in coco_manifest['annotations']:
+                images_by_id[annotation['image_id']].labels.append(annotation['caption'])
+            images = [x for x in images_by_id.values()]
+            images.sort(key=lambda x: x.id)
```
Comment (shonohs) [inline vision_datasets/common/data_manifest.py]: why don't we keep the original order? sorting seems to be unnecessary

File: vision_datasets/common/data_manifest.py [hunk 12]
```diff
@@ -447,6 +452,14 @@ def create_dataset_manifest(coco_file_path_or_url, data_type, container_sas_or_r
 
         file_reader.close()
 
+        if data_type == DatasetTypes.IMCAP:
+            images_by_id = {img['id']: ImageDataManifest(img['id'], get_full_sas_or_path(img['file_name']), None, None, []) for img in coco_manifest['images']}
```
Comment (pj-ms) [inline vision_datasets/common/data_manifest.py]: btw, for image width and height. Those are two typical info supported by COCO, not specific to caption or classification, or sth. It is ok to allow for missing values in caption case, but can we respect and take the info, if it is provided? Thanks

File: README.md [hunk 13]
```diff
@@ -18,7 +18,7 @@ This repo
 
 `ManifestDataset` is able to load the data from all three kinds of paths. Both 1. and 2. are good for training, as they access data from local disk while the 3rd one is good for data exploration, if you have the data in azure storage.
 
-Currently, three basic types of data are supported: `classification_multilabel`, `classification_multiclass`, and `object_detection`. `multitask` type is a composition type, where one set of images has multiple sets of annotations available for different tasks, where each task can be of the three basic types.
+Currently, four basic types of data are supported: `classification_multilabel`, `classification_multiclass`, `object_detection`, `image_caption`. `multitask` type is a composition type, where one set of images has multiple sets of annotations available for different tasks, where each task can be of the three basic types.
```
Comment (pj-ms) [inline README.md]: where each task can be of any basic type except for image_caption.

File: tests/test_create_dataset_manifest.py [hunk 14]
```diff
@@ -300,6 +300,47 @@ def test_object_detection_bbox_format_LTRB(self):
             self.assertEqual(dataset_manifest.images[0].labels, [[0, 10, 10, 100, 100]])
             self.assertEqual(dataset_manifest.images[1].labels, [[0, 100, 100, 200, 200], [1, 20, 20, 200, 200]])
 
+    def test_image_caption(self):
```
Comment (pj-ms) [inline tests/test_create_dataset_manifest.py:303]: can we add a test for multitask case where one task is caption and one task is classification? Thanks

File: vision_datasets/common/data_manifest.py (hunk 12 above)
Comment (msxingpeng) [reply vision_datasets/common/data_manifest.py]: The ground truth file that Xiaowei has shared with us does not contain the image width and height info. The width and height info are missing in the gt file, unless we want to add these info to the ground truth file.

Comment (pj-ms) [reply vision_datasets/common/data_manifest.py]: Not just Xiaowei is working on captioning...there could be people who works with captioning that do have width and height.

Comment (pj-ms) [reply vision_datasets/common/data_manifest.py]: "It is ok to allow for missing values in caption case, but can we respect and take the info, **if it is provided**?"

What I mean is 
    _respecting the info, if it is provided_
not
   _always requires width and height_

**Comment by pj-ms** [comment]
Could you bump the pkg version in setup.py

File: README.md [hunk 15]
```diff
@@ -18,7 +18,7 @@ This repo
 
 `ManifestDataset` is able to load the data from all three kinds of paths. Both 1. and 2. are good for training, as they access data from local disk while the 3rd one is good for data exploration, if you have the data in azure storage.
 
-Currently, three basic types of data are supported: `classification_multilabel`, `classification_multiclass`, and `object_detection`. `multitask` type is a composition type, where one set of images has multiple sets of annotations available for different tasks, where each task can be of the three basic types.
+Currently, four basic types of data are supported: `classification_multilabel`, `classification_multiclass`, `object_detection`, `image_caption`. `multitask` type is a composition type, where one set of images has multiple sets of annotations available for different tasks, where each task can be of any basic type except for `image_caption`.
```
Comment (pj-ms) [inline README.md]: Sorry for revisiting here. As we have a test making sure multitask can work for IC and captioning, we might want to update it to "where each task can be of any basic type".

File: vision_datasets/common/data_manifest.py (hunk 12 above)
Comment (msxingpeng) [reply vision_datasets/common/data_manifest.py]: I see. Have added a logic to check whether width and height are provided:

```python
            for img in coco_manifest['images']:
                image_width = None if 'width' not in img else img['width']
                image_height = None if 'height' not in img else img['height']
                images_by_id[img['id']] = ImageDataManifest(img['id'], get_full_sas_or_path(img['file_name']), image_width, image_height, [])

```

**Review by pj-ms — APPROVED** [review]
